### DISCUSSION TITLE
Implicit expression

### DISCUSSION THREAD

**Discussion by vincent-herlemont** [pest-parser/pest#761] [Q&A] [answered]
I made a tiny query language and I want to make an implicit boolean expression optional:
Example: `a && b` == `a a`, the user should have the possibility to write these two expressions.

Here the `.pest` file which describes the language:
```bash
program = _{ SOI ~ implicit ~ EOI  }
implicit= ${ or ~ (WHITESPACE+ ~ or )* }

or  = !{ and ~ (or_op ~ and)+ | and }
and = { comp ~ (and_op ~ comp)+ | comp }
comp = { array ~ eq_op ~ array | array }

array = ${ term }

term = _{ ASCII_ALPHANUMERIC+ }
or_op = { "||" }
and_op = { "&&" }
eq_op = { "=" }
WHITESPACE = _{ " " | "\t" | NEWLINE }
```

Example:
- `a && b` :heavy_check_mark: 
- `a a` :x: 
    ```
     --> 1:3
      |
    1 | a a
      |   ^---
      |
      = expected EOI, or_op, and_op, or eq_op
    ```

I don't understand why the implicit condition (`implicit= ${ or ~ (WHITESPACE+ ~ or )* }`) doesn't work in this case. Below I have simplified the syntax to test only the implicit mechanism:
```bash
program = _{ SOI ~ implicit ~ EOI  }
implicit= ${ term ~ (WHITESPACE+ ~ term )* }
term = _{ ASCII_ALPHANUMERIC+ }
WHITESPACE = _{ " " | "\t" | NEWLINE }
```

In this case the program `a a` works with sucess :heavy_check_mark: 

---

Context:
```
pest = "2.5.1"
pest_derive = "2.5.1"
```

**Comment by tomtau** [comment]
From a quick look at the grammar, I can't see what rules it should follow to just get the implicit option (it seems it'll always end with one of those explicit operator options?). Have you tried stepping through the parsing process using the CLI debugger: https://github.com/pest-parser/pest/discussions/739 ?

**Reply by vincent-herlemont** [reply]
@tomtau Thank you foryour answer. Thank you to inform me for the debugger, here his output:

```bash
> r program
 --> 1:1
  |
1 | a a
  | ^---
  |
  = parsing implicit
> c
 --> 1:1
  |
1 | a a
  | ^---
  |
  = parsing or
> c
 --> 1:1
  |
1 | a a
  | ^---
  |
  = parsing and
> c
 --> 1:1
  |
1 | a a
  | ^---
  |
  = parsing comp
> c
 --> 1:1
  |
1 | a a
  | ^---
  |
  = parsing array
> c
 --> 1:3
  |
1 | a a
  |   ^---
  |
  = expected "EOI", "and_op", "eq_op", or "or_op"
```

But the `*` (matches e zero or more times) of `implicit= ${ or ~ (WHITESPACE+ ~ or )* }` shouldn't authorize to access to the next `term` without `or_op` and `and_op`?

--- 

Other things, if I downgrade pest to `2.1.3` the above scenario works. I don't know if it's a regression or it's a fix for now, I think that I will open an issue.

**Comment by vincent-herlemont** [comment]
Related issue: https://github.com/pest-parser/pest/issues/762

**Comment by tomtau** [comment, marked as answer]
v2.5.3 was just published: https://github.com/pest-parser/pest/discussions/767 it should have a fix for that optimizer issue.

**Reply by JMLX42** [reply]
@tomtau fix confirmed!
